lib/Chunk.js

```diff
@@ -121,124 +117,136 @@ class Chunk {
 	}
 
 	remove(reason) {
-		this.modules.slice().forEach(m => {
-			m.removeChunk(this);
-		}, this);
-		this.parents.forEach(c => {
-			const idx = c.chunks.indexOf(this);
-			if(idx >= 0) {
-				c.chunks.splice(idx, 1);
-			}
-			this.chunks.forEach(cc => {
-				cc.addParent(c);
-			});
-		}, this);
-		this.chunks.forEach(c => {
-			const idx = c.parents.indexOf(this);
+		// cleanup modules
+		this.modules.slice().forEach(module => {
+			module.removeChunk(this);
+		});
+
+		// cleanup parents
+		this.parents.forEach(parentChunk => {
+			// remove this chunk from its parents
+			const idx = parentChunk.chunks.indexOf(this);
 			if(idx >= 0) {
-				c.parents.splice(idx, 1);
+				parentChunk.chunks.splice(idx, 1);
 			}
-			this.parents.forEach(cc => {
-				cc.addChunk(c);
+
+			// cleanup "sub chunks"
+			this.chunks.forEach(chunk => {
+				/**
+				 * remove this chunk as "intermediary" and connect
+				 * it "sub chunks" and parents directly
+				 */
+				// add parent to each "sub chunk"
+				chunk.addParent(parentChunk);
+				// add "sub chunk" to parent
+				parentChunk.addChunk(chunk);
```

Comment:
I think it happens if you extract an empty commons chunk (has no parent) and it is later removed by optimization because it's empty.

For the performance part you are correct. Could you just move the 4 lines below this comment into an new loop? So it's correct again.